### PR TITLE
MIPS porting (preliminary)

### PR DESCRIPTION
**Pull request by Harry-Chen**
* Port rCore to `mipsel` architecture
* Runs on qemu malta machine with uart and PCI (enumeration only) support
* Refactor some drivers
* Successfully runs most user apps

### CODE REVIEW

File: bootloader/src/arch/mipsel/boot.S [hunk 1]
```diff
@@ -0,0 +1,20 @@
+.section .text.boot
+.globl _start
+
+_start:
+    # read cpu affinity, start core 0, halt the rest
```
Comment (jiegec) [inline bootloader/src/arch/mipsel/boot.S]: Waiting for multicore support

Comment (Harry-Chen) [reply bootloader/src/arch/mipsel/boot.S]: not now

File: kernel/src/arch/mipsel/atomic.patch [hunk 2]
```diff
@@ -0,0 +1,72 @@
+--- atomic_backup.rs	2018-10-06 19:59:14.000000000 +0800
```
Comment (jiegec) [inline kernel/src/arch/mipsel/atomic.patch]: maybe not needed anymore in misp?

File: kernel/src/arch/mipsel/board/malta/mod.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        status.enable_hard_int2();
+        cp0::status::write(status);
+    }
+    println!("Hello QEMU Malta!");
+}
+
+/// Initialize other board drivers
+pub fn init_driver() {
+    // TODO: add possibly more drivers
+    vga::init(0xbbe00000, 0xb2050000, 800, 600);
```
Comment (jiegec) [inline kernel/src/arch/mipsel/board/malta/mod.rs]: can we find these address in dtb?

File: kernel/src/arch/mipsel/boot/entry.S [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+	.set noreorder
+    .section .text.entry
+    .globl _start
+	.extern _root_page_table_buffer
+	.extern _cur_kstack_ptr
+
+_start:
+    # setup stack and gp
+	la    sp, bootstacktop
+    la    gp, _gp
```
Comment (jiegec) [inline kernel/src/arch/mipsel/boot/entry.S]: Mixed tab and space?

File: kernel/src/arch/mipsel/board/malta/mod.rs (hunk 3 above)
Comment (Harry-Chen) [reply kernel/src/arch/mipsel/board/malta/mod.rs]: NO. VGA is needed by framebuffer driver and video console, which initializes BEFORE the driver framework.

Also this is a board specific file, I don't see any problem with hardcoding addresses.

**Comment by jiegec** [comment]
LGTM

File: kernel/src/arch/mipsel/boot/entry.S (hunk 4 above)
Comment (Harry-Chen) [reply kernel/src/arch/mipsel/boot/entry.S]: Will fix

File: kernel/src/arch/mipsel/atomic.patch (hunk 2 above)
Comment (Harry-Chen) [reply kernel/src/arch/mipsel/atomic.patch]: Will delete

File: kernel/src/arch/mipsel/board/malta/mod.rs (hunk 3 above)
Comment (Harry-Chen) [reply kernel/src/arch/mipsel/board/malta/mod.rs]: Maybe a comprehensive refactor of the driver framework can solve this.

**Comment by Harry-Chen** [comment]
RV32 and RV64 tests failed to pass. However, it does not seems to be our problem. 

Fixed some minor problems pointed out by jiegec and added mipsel to CI.

File: bootloader/src/arch/mipsel/boot.S (hunk 1 above)
Comment (Harry-Chen) [reply bootloader/src/arch/mipsel/boot.S]: will add multicore support, maybe later this month

**Comment by wangrunji0408** [comment]
I'm reviewing...
Orz first for your great contribution.

A personal suggestion: 
It's unfriendly to macOS developers now. 😂
Is there a binary release of `mipsel-linux-musln32-gcc` on macOS?
Or could we get rid of the gcc toolchain for kernel?
It seems that only assembly code depends on it.
I think `.macro` or `.equ` in asm may achieve the same effect.

And it's better to run `cargo fmt` before the final merge.

**Comment by Harry-Chen** [comment]
There seems no .macro support in MIPS assembly, and I could not figure out why. But to format the code is a good suggestion.

**Comment by Harry-Chen** [comment]
But I can change to any version of `gcc`, since we only use it for preprocessing.

**Comment by wangrunji0408** [comment]
A crazy idea: preprocess the asm using `build.rs`

**Comment by Harry-Chen** [comment]
> A crazy idea: preprocess the asm using `build.rs`

Might work. But I have now changed to `gcc`, which I believe adds no difficulty to any developer

File: kernel/src/arch/mipsel/compiler_rt.rs [hunk 5]
```diff
@@ -0,0 +1,8 @@
+//! Workaround for missing compiler-builtin symbols
+//!
+//! [atomic](http://llvm.org/docs/Atomics.html#libcalls-atomic)
+
+#[no_mangle]
+pub extern "C" fn abort() {
```
Comment (wangrunji0408) [inline kernel/src/arch/mipsel/compiler_rt.rs]: This can be removed on mips.

File: kernel/src/arch/mipsel/context.rs [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+    pub s7: usize,
+    pub t8: usize,
+    pub t9: usize,
+    pub k0: usize,
+    pub k1: usize,
+    pub gp: usize,
+    pub sp: usize,
+    pub fp: usize,
+    pub ra: usize,
+    /// Reserve space for hartid
```
Comment (wangrunji0408) [inline kernel/src/arch/mipsel/context.rs]: No longer needed since we can get core id from `cp0.ebase`

**Pull request merged**